### PR TITLE
Deploy to development in pipelines

### PR DESCRIPTION
**Pull request by pezholio**
This uses the Azure Pipeline to deploy as well as run the tests if the branch is master. This keeps the two things tightly coupled, and also allows us to keep our deployment steps tightly coupled to our code, rather than having them defined in Azure DevOps via a UI. This can also be adapted to deploy to production when we're ready to do so.

I've also put all our Docker-related tasks in a seperate bin script, so it's easier to follow, rather than having to parse a massive bunch of YAML.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -6,106 +6,58 @@ pool:
...
+            env:
+              pass: $(dockerPassword)
+            displayName: Login to DockerHub
+          - script: |
+              ./bin/build-docker-images "ref/heads/master" \
+                $(Build.SourceVersion) \
+                $(dockerRegistry)/$(dockerImageName) \
+                $(Build.BuildNumber)
+            displayName: Prepare and push Docker images
+      - job: publish_repository
```
Comment (robbiepaul) [inline azure-pipelines.yml]: Do we only want to publish the repository when merging to master?

File: bin/slack-alert [hunk 2]
```diff
@@ -2,11 +2,11 @@
 
 ENVIRONMENT_NAME=$1
 BUILD_NUMBER=$2
-RELEASE_ID=$3
+BUILD_ID=$3
 WEBHOOK_URL=$4
 RESULT=$5
 
-RELEASE_URL="https://dev.azure.com/dfe-ssp/S118-Teacher-Payments-Service/_releaseProgress?_a=release-pipeline-progress&releaseId=$RELEASE_ID"
+RELEASE_URL="https://dfe-ssp.visualstudio.com/S118-Teacher-Payments-Service/_build/results?buildId=$BUILD_ID&view=results"
```
[the remark this answers is not in the document]
Comment (pezholio) [reply bin/slack-alert]: The release happens as part of the build now (see https://dfe-ssp.visualstudio.com/S118-Teacher-Payments-Service/_build/results?buildId=50614&view=results for an example)

File: azure-pipelines.yml (hunk 1 above)
Comment (pezholio) [reply azure-pipelines.yml]: Yeah, this just publishes the artifact, which we need for deploy.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -6,106 +6,58 @@ pool:
...
+      - job: build_and_push_web_dependencies
+        dependsOn: run_tests
+        displayName: Build and push web dependencies
+        steps:
+          - script: docker login -u $(dockerId) -p $pass
+            env:
+              pass: $(dockerPassword)
+            displayName: Login to DockerHub
+          - script: |
+              ./bin/build-docker-images "ref/heads/master" \
```
[the remark this answers is not in the document]
Comment (pezholio) [reply azure-pipelines.yml]: Good call 👍

File: bin/slack-alert (hunk 2 above)
[the remark this answers is not in the document]
Comment (robbiepaul) [reply bin/slack-alert]: Ahh ok gotcha

File: azure-pipelines.yml (hunk 1 above)
Comment (robbiepaul) [reply azure-pipelines.yml]: It doesn't have a `condition` though, so won't it publish the artifact on non-master builds?

Comment (pezholio) [reply azure-pipelines.yml]: Oh gotcha. You're right 😄

**Comment by pezholio** [comment]
I'm gonna take a look at seperating out the caching and building of Docker images, they do seem like separate things

**Pull request closed without merging**